src/xrpld/app/misc/detail/LendingHelpers.cpp

```diff
@@ -0,0 +1,1643 @@
+    }
+
+    // Make sure the parts don't add up to too much
+    Number shortage = roundedPeriodicPayment - deltas.total();
+
+    XRPL_ASSERT_PARTS(
+        isRounded(asset, shortage, scale),
+        "ripple::detail::computePaymentComponents",
+        "shortage is rounded");
+
+    if (shortage < beast::zero)
+    {
+        Number excess = -shortage;
+
+        addressExcess(deltas, excess);
+
+        shortage = -excess;
+    }
+    // The shortage should never be negative, which indicates that the
+    // parts are trying to take more than the whole payment. The
+    // shortage may be positive, which indicates that we're not going to
+    // take the whole payment amount.
+    XRPL_ASSERT_PARTS(
+        shortage >= beast::zero,
+        "ripple::detail::computePaymentComponents",
+        "no shortage or excess");
+
+    XRPL_ASSERT_PARTS(
+        deltas.total() ==
+            deltas.principal + deltas.interest + deltas.managementFee,
+        "ripple::detail::computePaymentComponents",
+        "total value adds up");
+
+    XRPL_ASSERT_PARTS(
+        deltas.principal >= beast::zero &&
+            deltas.principal <= currentLedgerState.principalOutstanding,
+        "ripple::detail::computePaymentComponents",
+        "valid principal result");
+    XRPL_ASSERT_PARTS(
+        deltas.interest >= beast::zero &&
+            deltas.interest <= currentLedgerState.interestDue,
+        "ripple::detail::computePaymentComponents",
+        "valid interest result");
+    XRPL_ASSERT_PARTS(
+        deltas.managementFee >= beast::zero &&
+            deltas.managementFee <= currentLedgerState.managementFeeDue,
+        "ripple::detail::computePaymentComponents",
+        "valid fee result");
+
+    XRPL_ASSERT_PARTS(
+        deltas.principal + deltas.interest + deltas.managementFee > beast::zero,
+        "ripple::detail::computePaymentComponents",
+        "payment parts add to payment");
+
+    return PaymentComponents{
+        // As a final safety check, ensure the value is non-negative, and won't
+        // make the corresponding item negative
+        .trackedValueDelta = std::clamp(
+            deltas.total(), numZero, currentLedgerState.valueOutstanding),
+        .trackedPrincipalDelta = std::clamp(
+            deltas.principal, numZero, currentLedgerState.principalOutstanding),
+        .trackedManagementFeeDelta = std::clamp(
+            deltas.managementFee, numZero, currentLedgerState.managementFeeDue),
+    };
+}
+
+ExtendedPaymentComponents
+computeOverpaymentComponents(
+    Asset const& asset,
+    int32_t const loanScale,
+    Number const& overpayment,
+    TenthBips32 const overpaymentInterestRate,
+    TenthBips32 const overpaymentFeeRate,
+    TenthBips16 const managementFeeRate)
+{
+    XRPL_ASSERT(
+        overpayment > 0 && isRounded(asset, overpayment, loanScale),
+        "ripple::detail::computeOverpaymentComponents : valid overpayment "
+        "amount");
+
+    Number const fee = roundToAsset(
```

Comment:
Perhaps `overpaymentFee` is a better name.

Reply:
> Perhaps `overpaymentFee` is a better name.

Fixed in 1ba010031d560c485f05e5e74fdbb0c4d2e171ca